src/binder/bind_node_visitor.cpp

```diff
@@ -354,25 +356,60 @@ void BindNodeVisitor::Visit(parser::InsertStatement *node, parser::ParseResult *
             throw BINDER_EXCEPTION("Mismatch in number of insert columns and number of insert values.");
           }
         }
-        // Test that the column values are of the right type.
-        for (size_t i = 0; i < num_values; ++i) {
-          // TODO(WAN): handle additional cases, possibly think about calling DeriveReturnValueType
-          //  so that we handle (1+2) type of expressions
-          //  ADDENDUM. So I thought DeriveReturnValueType would actually derive the return value type.
-          //  This appears to be a bad assumption. We should rename it to DeriveReturnValueTypeForAggregates()
-          //  or else fix up any other codepaths. I've currently fixed it for ConstantValueExpression.
-          auto expr = values[i];
-          auto ret_type = expr->GetReturnValueType();
-          auto expected_ret_type = table_schema.GetColumn(i).Type();
-
-          auto is_cast_expression = expr->GetExpressionType() == parser::ExpressionType::OPERATOR_CAST;
+
+        std::vector<std::pair<catalog::Schema::Column, common::ManagedPointer<parser::AbstractExpression>>> cols;
+        if (num_insert_columns == 0) {  //already ordered
+          for (int i = 0; i < num_values; i++) {
+            std::pair<catalog::Schema::Column, common::ManagedPointer<parser::AbstractExpression>> pair =
+                std::make_pair(table_schema.GetColumns()[i], values[i]);
+            cols.push_back(pair);
+          }
+        } else {
+          for (int i = 0; i < num_insert_columns; i++) {
```

Comment:
Let's do it as we discussed after meeting.

1. Go through the schema columns.
2. For each column, if the name of the column is in the InsertColumns, then insert the corresponding value.
3. If not, if the column has default value, insert that.
4. If not, if the column is nullable, insert the corresponding null value.
5. If not, throw a binder exception.

You also need to check if they do not give you insert columns.
If not, maybe you can copy the column names out in schema order from the schema.
It will be slower but the code will be cleaner and we can fix the slowness if we see it being a bottleneck.